src/controller/DeviceUnpair.cpp

```diff
@@ -0,0 +1,187 @@
+
+    OperationalCredentials::Commands::RemoveFabric::Type request;
+    request.fabricIndex = fabricIndex;
+
+    OperationalCredentialsCluster cluster(*proxy->GetExchangeManager(), proxy->GetSecureSession().Value(), 0);
+
+    cluster.InvokeCommand(request, this, OnRemoveFabric, OnCommandFailure);
+}
+
+void DeviceUnpair::FinishUnpairDevice(CHIP_ERROR err)
+{
+    mAttributeCache         = nullptr;
+    mReadClient             = nullptr;
+    mProxy                  = nullptr;
+    mOperationalDeviceProxy = nullptr;
+
+    if (mCallback != nullptr)
+    {
+        mCallback->OnDeviceUnpair(mRemoteDeviceId, err);
+    }
+}
+
+void DeviceUnpair::OnDeviceConnectedFn(void * context, Messaging::ExchangeManager & exchangeMgr, SessionHandle & sessionHandle)
+{
+    DeviceUnpair * deviceUnpair = static_cast<DeviceUnpair *>(context);
+    VerifyOrReturn(deviceUnpair != nullptr, ChipLogProgress(Controller, "Device connected callback with null context. Ignoring"));
+
+    auto proxy                            = Platform::MakeShared<OperationalDeviceProxy>(&exchangeMgr, sessionHandle);
+    deviceUnpair->mOperationalDeviceProxy = std::move(proxy);
+    deviceUnpair->PerformDeviceUnpairStep(deviceUnpair->mOperationalDeviceProxy.get(), UnpairDeviceStage::kReadCurrentFabricIndex,
+                                          CHIP_NO_ERROR);
+}
+
+void DeviceUnpair::OnDeviceConnectionFailureFn(void * context, const ScopedNodeId & peerId, CHIP_ERROR err)
+{
+    ChipLogProgress(Controller, "OnDeviceConnectionFailureFn %s", err.AsString());
+
+    DeviceUnpair * deviceUnpair = static_cast<DeviceUnpair *>(context);
+    VerifyOrReturn(deviceUnpair != nullptr, ChipLogProgress(Controller, "Device connected callback with null context. Ignoring"));
+
+    deviceUnpair->PerformDeviceUnpairStep(nullptr, UnpairDeviceStage::kUnpairError, err);
+}
+
+void DeviceUnpair::OnDone(app::ReadClient * apReadClient)
+{
+    FabricIndex fabricIndex = kUndefinedFabricIndex;
+    CHIP_ERROR err          = mAttributeCache->ForEachAttribute(
+        OperationalCredentials::Id, [this, &fabricIndex](const app::ConcreteAttributePath & path) {
+            if (path.mAttributeId != OperationalCredentials::Attributes::CurrentFabricIndex::Id)
+            {
+                // Continue on
+                return CHIP_NO_ERROR;
+            }
+
+            switch (path.mAttributeId)
+            {
+            case OperationalCredentials::Attributes::CurrentFabricIndex::Id:
+                return this->mAttributeCache->Get<OperationalCredentials::Attributes::CurrentFabricIndex::TypeInfo>(path,
+                                                                                                                    fabricIndex);
+            default:
+                return CHIP_NO_ERROR;
+            }
+        });
+    mFabricIndex = fabricIndex;
+    PerformDeviceUnpairStep(mProxy, UnpairDeviceStage::kSendRemoveFabric, err);
```

Comment:
Note that at this point we might not have a useful secure session; we should really re-get that for each async step....